src/guards/slice_memory_guard.rs

```diff
@@ -50,6 +51,53 @@ impl<'a, T> SliceMemoryGuard<'a, T> {
         }
         SliceMemoryGuard { memory }
     }
+
+    /// Initialize memory guard using given iterator.
+    /// Automatically shrink's memory to given items' count.
+    /// `Ok(guard)` will be returned in this case.
+    ///
+    /// If items' count is too large to place in memory, moves it into new `Vec` and continue collecting into it.
+    /// `Err(vec)` will be returned in this case.
+    #[inline]
+    pub unsafe fn new_from_iter(memory: &'a mut [MaybeUninit<T>], mut iter: impl Iterator<Item=T>) -> Result<Self, Vec<T>> {
+        // Fulfilling placed memory
+        for (index, item) in memory.into_iter().enumerate() {
+            match iter.next() {
+                // While iterator returns new value, write it
+                Some(value) => {
+                    write(item.as_mut_ptr(), value);
+                }
+                // When it returns None then slicing memory and returning the guard
+                None => {
+                    return Ok(SliceMemoryGuard {
+                        memory: &mut memory[0..index],
+                    });
+                }
+            }
+        }
+
+        if let Some(next_item) = iter.next() {
+            let memory = transmute::<&'a mut [MaybeUninit<T>], &'a mut [T]>(memory);
```

Comment:
nit: you don't need to specify this source type if it's already known, i.e. `transmute::<_, Xxx>` should do the trick. Alternatively, even `let memory: &'a mut [T] = transmute(memory)`

Reply:
`transmute` is pretty dangerous for mistakes. So I prefer to fill all parameters explicitly.